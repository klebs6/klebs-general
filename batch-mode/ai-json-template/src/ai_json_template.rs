// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}

/// Extended trait that demands both a normal `AiJsonTemplate` and
/// a `...Justification` structure for *nested* usage.
pub trait AiJsonTemplateWithJustification: AiJsonTemplate {
    fn to_template_with_justification() -> serde_json::Value;
}
