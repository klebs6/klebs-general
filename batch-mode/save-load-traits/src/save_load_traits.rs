// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
