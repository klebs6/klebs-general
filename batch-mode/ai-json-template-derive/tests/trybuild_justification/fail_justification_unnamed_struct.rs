// ---------------- [ File: ai-json-template-derive/tests/trybuild_justification/fail_justification_unnamed_struct.rs ]
// tests/trybuild_justification/fail_justification_unnamed_struct.rs
//
// A tuple struct deriving AiJsonTemplateWithJustification => must fail
// because we only support named fields.

#![allow(dead_code)]
#![allow(unused_imports)]

use ai_json_template::*;
use ai_json_template_derive::*;
use serde::*;
use getset::*;
use derive_builder::Builder;
use save_load_derive::*;
use save_load_traits::*;
use tracing::*;

#[derive(SaveLoad, Debug, Clone, Serialize, Deserialize, Builder)]
#[derive(AiJsonTemplate, AiJsonTemplateWithJustification)]
struct BasicUnnamed(u32, String);

fn main() {}
