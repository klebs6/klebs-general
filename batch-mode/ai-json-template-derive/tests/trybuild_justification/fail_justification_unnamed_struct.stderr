error: Unsupported shape `unnamed fields`. Expected named fields.
  --> tests/trybuild_justification/fail_justification_unnamed_struct.rs:18:58
   |
18 | #[derive(SaveLoad, Debug, Clone, Serialize, Deserialize, Builder)]
   |                                                          ^^^^^^^
   |
   = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: AiJsonTemplate derive only supports named fields for structs (or handle your tuple struct in the enum logic).
  --> tests/trybuild_justification/fail_justification_unnamed_struct.rs:20:1
   |
20 | struct BasicUnnamed(u32, String);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: AiJsonTemplateWithJustification only supports named fields for structs.
  --> tests/trybuild_justification/fail_justification_unnamed_struct.rs:20:1
   |
20 | struct BasicUnnamed(u32, String);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
